### PR TITLE
feat(rh-tile): add rh-tile

### PR DESCRIPTION
**Pull request by nikkimk**
## What I did

1. Created `rh-tile`.
2. Created `rh-tile-group`.
3. Documentation on both.

## Testing Instructions

1. See [DP Demo](https://deploy-preview-1239--red-hat-design-system.netlify.app/elements/tile/demo/color-context/)
2. See [DP Docs](https://deploy-preview-1239--red-hat-design-system.netlify.app/elements/tile/)

## Notes to Reviewers
Mockup: https://xd.adobe.com/view/925ed24c-3ddf-4111-bacf-2a42e5249f7a-c70d/

### CODE REVIEW

**Review by marionnegp — CHANGES_REQUESTED** [review]
@eyevana, let me know if you need help with some of this! I feel like this is quite a bit, so I can work on edits to the tile docs tomorrow.

- [ ] Footer slot shouldn't have a bottom margin and should look more aligned with the arrow icon.
- [ ] This version in the demo might just be there for testing, but it shouldn't exist because it has a link in the body text. Screenshot 2023-10-12 at 8 42 40 PM

Link tile, default size:
- [ ] space between image and icon should be `--rh-space-2xl`
- [ ] space between icons and text should be `--rh-space-2xl`
- [ ] space between heading and body text should be`--rh-space-lg`
- [ ] With the 16px bottom margin removed from footer slot, the bottom padding of the tile needs to be `--rh-space-2xl`. 

Link tile, compact size:
- [ ] Padding for the whole tile should be `--rh-space-xl`
- [x] For the full-width image version, margin below image should be `--rh-space-xl`
- [x] Space between heading and body text should be `--rh-space-lg`
- [x] Remove title slot

Selectable tile, checkboxes:
- [x] Space between heading and body text should be `--rh-space-lg`
- [x] Remove title slot

Selectable tile, radio buttons:
- [x] Space between heading and body text should be `--rh-space-lg`
- [x] Remove title slot

Hover, focus, and active states:
- [ ] selectable tiles with checkbox and radio buttons need to be underlined
- [x] For all tile variants (except for desaturated heading variants), the heading and arrow color need to change to `--rh-color-interactive-blue-darkest` (light theme) or `--rh-color-interactive-blue-lightest` (dark theme)

Tile group:
- [x] Need space between tile rows.  
Screenshot 2023-10-12 at 8 49 45 PM

File: .changeset/dry-pugs-melt.md
Comment (bennypowers) [inline .changeset/dry-pugs-melt.md:1]: formatting. see https://github.com/RedHat-UX/red-hat-design-system/releases/tag/v1.1.0

File: .changeset/dry-pugs-melt.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  <rh-tile>
+    <img slot="image" src="//fakeimg.pl/296x50" alt="296 X 50 placeholder image">
+    <div slot="title">Title</div>
+    <h2 slot="headline"><a href="#top">Link</a></h2>
+    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
+    <div slot="footer">Suspendisse eu turpis elementum</div>
+  </rh-tile>
+```
+
+Added `<rh-tile-group>`.
```
Comment (bennypowers) [inline .changeset/dry-pugs-melt.md:19]: if this doesn't have it's own top-level directory, don't include it here

if you do break this out into a top-level directly, split it into a separate changeset

if you keep it within `elements/rh-tile/rh-tile-group.js`, i suggest replacing the above snippet with this one, but keepin the prose the same.

File: elements/rh-tile/demo/rh-tile.html [hunk 2]
```diff
@@ -0,0 +1,239 @@
+<link rel="stylesheet" href="demo.css">
+<link rel="stylesheet" href="../rh-tile-lightdom.css">
+<script type="module" src="rh-tile.js"></script>
+
+<script type="module">
+  import '@rhds/elements/lib/elements/rh-context-picker/rh-context-picker.js';
+  import '@rhds/elements/lib/elements/rh-context-provider/rh-context-provider.js';
+</script>
+
+<rh-context-picker target="demo-container"></rh-context-picker>
```
Comment (bennypowers) [inline elements/rh-tile/demo/rh-tile.html:10]: split into separate files. context-picker etc should only be in color-context.htmj

File: elements/rh-tile/demo/rh-tile.js
Comment (bennypowers) [inline elements/rh-tile/demo/rh-tile.js:1]: inline these imports into each demo. remove the last two statements.

File: elements/rh-tile/docs/10-style.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+           alt="Default link tile with numbers pointing to locations of an image, an icon, text, and a footer",
+           src="../link-tile-anatomy.png" %}
+ 
+1) Image
+2) Icon
+3) Title
+4) Heading
+5) Body text
+6) Footer
+{.example-notes}
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:17]: this might need 2 spaces of indent. verify that it renders correctly on ux.

File: elements/rh-tile/docs/10-style.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+### Sizes
+
+A link tile is available in Default and Compact sizes. A selectable tile has only one size which is based on the size of a compact tile.
+
+{% example palette="light",
+           alt="Examples of a link tile, compact tile, and selectable tile to show size differences",
+           src="../tile-sizes.png" %}
+
+<!---
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:37]: remove

File: elements/rh-tile/docs/10-style.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    
+{% example palette="light",
+        alt="Light theme tiles use a white background, blue or black heading, black text, and a blue arrow icon",
+        src="../tile-light-theme.png" %}
+
+{% example palette="darkest",
+        alt="Dark theme tiles use a dark gray background, blue or white heading, white text, and a light blue arrow icon",
+        src="../tile-dark-theme.png" %}
+
+<!---
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:73]: remove

File: elements/rh-tile/docs/10-style.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        alt="Link tiles that have full-width and default image sizes with spacers showing padding and margins",
+        src="../space-link-tile-with-image.png" %}
+
+### Selectable tile
+
+{% example palette="light",
+        alt="Selectable tile with spacers showing padding and margins",
+        src="../space-selectable-tile.png" %}
+
+<!---
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:131]: remove

File: elements/rh-tile/docs/10-style.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+{% example palette="light",
+        alt="On hover, light theme tiles have a light gray background, an underlined (and sometimes darker blue) heading, a darker blue arrow icon ",
+        src="../tile-states-hover-light-theme.png" %}
+
+{% example palette="darkest",
+        alt="On hover, dark theme tiles have a lighter gray background, an underlined (and sometimes lighter blue) heading, a lighter blue arrow icon ",
+        src="../tile-states-hover-dark-theme.png" %}
+
+<!---
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:153]: remove

File: elements/rh-tile/docs/10-style.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+{% example palette="light",
+        alt="Focused light theme tiles have a blue focus ring and use hover state styling",
+        src="../tile-states-focus-light-theme.png" %}
+
+{% example palette="darkest",
+        alt="Focused dark theme tiles have a light blue focus ring and use hover state styling",
+        src="../tile-states-focus-dark-theme.png" %}
+
+<!---
```
Comment (bennypowers) [inline elements/rh-tile/docs/10-style.md:183]: remove

File: elements/rh-tile/docs/20-guidelines.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+The recommended character counts below include spaces. Line counts are based on a default link tile at minimum width.
+
+| Element                  | Character count  | Line count |
+| -------------------------| ---------------- | ---------- |
+| Title text               | 20               | 1          |
+| Heading text             | 64               | 3          |
+| Body text                | 160              | 7          |
+| Footer text              | 25               | 1          |
+
+{.full-width .col-111}
```
Comment (bennypowers) [inline elements/rh-tile/docs/20-guidelines.md:66]: verifies this renders as intended

File: elements/rh-tile/docs/30-code.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{% renderInstallation %}{% endrenderInstallation %}
+
+{% band header="Usage" %}
+
```
Comment (bennypowers) [inline elements/rh-tile/docs/30-code.md:4]:
```suggestion
```

File: elements/rh-tile/docs/40-accessibility.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (bennypowers) [inline elements/rh-tile/docs/40-accessibility.md:1]:
```suggestion
```

File: elements/rh-tile/rh-tile-group.css [hunk 12]
```diff
@@ -0,0 +1,8 @@
+:host([aria-disabled="true"]) {
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile-group.css:1]: we should prefer our own `disabled` attr, not expose `aria-` as public apis

File: elements/rh-tile/rh-tile-group.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+ */
+@customElement('rh-tile-group')
+export class RhTileGroup extends LitElement {
+  static readonly styles = [styles];
+
+  /**
+   * whether tile interaction is disabled but retains ability to be read by screen readers;
+   * preferred method of disabling instead of `disabled`
+   */
+  @property({ reflect: true, attribute: 'aria-disabled', type: String }) ariaDisabled = 'false';
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile-group.ts:26]: we should prefer to expose our own `disabled` and use `aria` stuff as internal apis

File: elements/rh-tile/rh-tile-group.ts [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+  /**
+   * if tile is checkable, whether only on tile can be checked
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile-group.ts:29]:
```suggestion
   * if tile is checkable, whether only one tile can be checked
```

what about `<rh-tile checkable="radio">` instead of `<rh-tile checkable radio>`?

File: elements/rh-tile/rh-tile.ts [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+export class RhTile extends LitElement {
+  static readonly styles = [styles];
+
+  private static readonly _disabledIcon = html`<svg xmlns="http://www.w3.org/2000/svg" width="48" height="48" viewBox="0 0 48 48"><g id="uuid-0fd9e805-a455-40ef-9171-f2f334832bf2"><rect width="48" height="48" fill="none"/></g><g id="uuid-48f9e284-0601-4fcd-bbe7-8b444234ac6c"><path d="m24,7c-9.37,0-17,7.63-17,17s7.63,17,17,17,17-7.63,17-17S33.37,7,24,7Zm15,17c0,3.52-1.23,6.76-3.27,9.32L14.68,12.27c2.56-2.04,5.8-3.27,9.32-3.27,8.27,0,15,6.73,15,15Zm-30,0c0-4.03,1.61-7.69,4.2-10.38l21.18,21.18c-2.7,2.6-6.35,4.2-10.38,4.2-8.27,0-15-6.73-15-15Z"/></g></svg>`;
+
+  /**
+   * whether tile interaction is disabled but retains ability to be read by screen readers;
+   * preferred method of disabling instead of `disabled`
+   */
+  @property({ reflect: true, attribute: 'aria-disabled', type: String }) ariaDisabled = 'false';
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile.ts:53]: see notes on tile group

File: elements/rh-tile/rh-tile.ts [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+   */
+  @property({ attribute: 'icon', type: String }) icon = false;
+
+  /**
+   * whether tile can be checked like a radio or checkbox:
+   * `false` (default) - tile behaves like a link;
+   * `true` - tile behaves like a checkbox unless it is part of an
+   * `rh-tile-group` with a `radio` type and more than one tile
+   */
+  @property({ attribute: 'checkable', type: Boolean }) checkable = false;
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile.ts:82]:
```suggestion
  @property({ type: Boolean }) checkable = false;
```

Comment (bennypowers) [inline elements/rh-tile/rh-tile.ts:82]:
```suggestion
  @property({
    converter: booleanEnumConverter({ values: ['radio'] })
  }) checkable = false;
```
consider making this boolean-or-value. 

tests:

```html
<rh-tile id=none></rh-tile>
<rh-tile id=bare interactive></rh-tile>
<rh-tile id=empty interactive=""></rh-tile>
<rh-tile id=silly interactive="silly"></rh-tile>
<rh-tile id=radio interactive="radio"></rh-tile>
<rh-tile id=case interactive="RaDio"></rh-tile>
<script>
  console.assert(none.interactive  === false); 
  console.assert(bare.interactive  === true); 
  console.assert(empty.interactive === true); 
  console.assert(silly.interactive === true); 
  console.assert(radio.interactive === 'radio');
  console.assert(case.interactive  === 'radio');
</script> 
```
if we don't already have a converter-constructor function for this in pfe-core or in `/lib`, now's the time, there are at least 2 other instances (perhaps in open PRs or in pfe, so please check with the team)

File: elements/rh-tile/test/rh-tile-group.spec.ts [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+        .to.be.an.accessible;
+    });
+
+    it('has radio buttons', async function() {
+      element = await createFixture<RhTileGroup>(tileGroup);
+      expect(tiles(element).length)
+        .to.equal(inputs(element).length);
+    });
+
+    describe('acts as a radio group', async function() {
```
Comment (bennypowers) [inline elements/rh-tile/test/rh-tile-group.spec.ts:93]:
```suggestion
    describe('as a radio group', async function() {
```

combine this with the next sibling describe, since they use the same template. inline the template

File: elements/rh-tile/test/rh-tile.spec.ts [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+    it('is accessible', async function() {
+      element = await createFixture<RhTile>(tile);
+      expect(element)
+        .to.be.an.accessible;
+    });
+  });
+
+  describe('checkable tile', function() {
+    const tile = html`<rh-tile checkable>
```
Comment (bennypowers) [inline elements/rh-tile/test/rh-tile.spec.ts:57]: put this and create fixture in beforeEach, `it` should as a general rule not be async, except for when you need to `await a11ySnapshot` or when you need to `await expect(element).to.be.accessible()`

File: elements/rh-tile/test/rh-tile.spec.ts [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+      <div slot="title">Title</div>
+      <h2 slot="headline"><a href="#top">Link</a></h2>
+      Lorem ipsum dolor sit amet, consectetur adipiscing elit.
+      <div slot="footer">Suspendisse eu turpis elementum</div>
+    </rh-tile>`;
+    let element: RhTile;
+    it('is accessible', async function() {
+      element = await createFixture<RhTile>(tile);
+      expect(element)
+        .to.be.an.accessible;
```
Comment (bennypowers) [inline elements/rh-tile/test/rh-tile.spec.ts:68]:
```suggestion
      await expect(element).to.be.accessible();
```

File: elements/rh-tile/test/rh-tile.spec.ts [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    await sendMouse({ type: 'click', position });
+  }
+
+  async function enter() {
+    await sendKeys({ press: 'Enter' });
+    await element.updateComplete;
+  }
+
+  function input(element, checked = false) {
+    return element?.shadowRoot?.querySelector(`input[type="checkbox"]${checked ? ':checked' : ''}`);
```
Comment (bennypowers) [inline elements/rh-tile/test/rh-tile.spec.ts:21]: instead of asserting on the shadow root, use a11yShapshot and assert against the ax tree

Comment (bennypowers) [reply elements/rh-tile/test/rh-tile.spec.ts:21]: and i think you can inline this

File: elements/rh-tile/test/rh-tile.spec.ts [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        element.focus();
+        await enter();
+        expect(element.checked)
+          .to.equal(true);
+
+        describe('clicking', async function() {
+          it('is unchecked', async function() {
+            element = await createFixture<RhTile>(tile);
+            element.checked = true;
+            await click(element);
```
Comment (bennypowers) [inline elements/rh-tile/test/rh-tile.spec.ts:89]: setup preliminary state in `describe/beforeEach` and perform actions in `beforeEach`. `it` should (usually synchronously) only assert on state

**Review by bennypowers — CHANGES_REQUESTED** [review]

File: elements/rh-tile/rh-tile-group.ts [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+  /** sets focus on active tile */
+  focus() {
+    (this.#tabindex?.activeItem || this.#tabindex.firstItem)?.focus();
+  }
+
+  /**
+   * programatically select a tile
+   * @param tile {RhTile | null | undefined} tile to select
+   */
+  selectTile(tile: RhTile | null | undefined) {
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile-group.ts:115]: we should probably settle on a nomenclature and call all of these `selectItem` etc, aligning names around RTIController in all element

**Review by marionnegp — CHANGES_REQUESTED** [review]
Tile element edits:
- [ ] There should be 32px between the image and title in link tiles
- [ ] The arrow color on desaturated link tiles should use `--rh-color-interactive-blue-darkest` in hover, focus, and active states.
- [ ] I don't think the focus indicator should change color.
- [x] Selectable tiles shouldn't have title or image slots.
- [x] Link tiles shouldn't have any other linked content outside of the heading (or image or whatever's getting the `<a>` tag). Selectable tiles shouldn't have any linked content. Those only can be checkboxes or radio buttons.

Tile docs:
- [x] Remove underline from hover, focus, and selected states in the docs.
- [x] Correct the focus state in keyboard interaction and focus order images

@nikkimk, I can update the docs.

File: elements/rh-tile/rh-tile-group.ts (hunk 22 above)
Comment (bennypowers) [reply elements/rh-tile/rh-tile-group.ts:115]: to clarify this should be cross referenced and changed before release in order to prevent breaking changes

**Review by bennypowers — APPROVED** [review]

**Comment by bennypowers** [comment]
Approval sent just now is a promissory note against the changes i requested

**Review by marionnegp — CHANGES_REQUESTED** [review]
@nikkimk, I checked off edits from previous reviews that were done. Thank you! Just a few more changes that I'm not seeing done in the demo.

- Default-sized link tiles with full-width images still show a 16px spacing between image and title.
Screenshot 2023-10-16 at 2 50 59 PM

- Focus state doesn't show darker blue heading and arrow or underlined heading for link tiles.
Screenshot 2023-10-16 at 2 52 05 PM

- Can you double check spacing for compact tile? You might need to add another custom variable to get that 16px.
Screenshot 2023-10-16 at 3 02 13 PM

**Comment by marionnegp** [comment]
@nikkimk,  woohoo! I got to check off the focus style change and the spacing between the title and heading.

Spacing for compact tile is still a little off. Hopefully it's not a caching issue on my end, but here's what it looks like for me. Still seeing a margins of 12px and 24px where it should be 16px in compact tile.
Screenshot 2023-10-16 at 3 41 35 PM

Screenshot 2023-10-16 at 3 40 55 PM

Screenshot 2023-10-16 at 3 39 30 PM

**Review by marionnegp — CHANGES_REQUESTED** [review]
@nikkimk 
- I'm getting some extra space between images and the heading on compact link after the 16px update. Should just be a space of 24px.

Screenshot 2023-10-16 at 4 29 12 PM

- Also, I think I overlooked the extra space between the image and an icon here in my last comment. I'm sorry!
Screenshot 2023-10-16 at 4 30 05 PM

**Review by marionnegp — CHANGES_REQUESTED** [review]
It looks like there's only 16px under the image for default and compact sizes now 😬.

Compact size:
Screenshot 2023-10-16 at 5 10 06 PM

Default size:
Screenshot 2023-10-16 at 5 09 37 PM

**Review by marionnegp — CHANGES_REQUESTED** [review]
Still showing the 16px space under images

**Review by marionnegp — CHANGES_REQUESTED** [review]
@nikkimk, the default-sized link tile looks great! Still seeing only 16px under the image for compact link tile.

**Review by marionnegp — APPROVED** [review]

File: elements/rh-tile/rh-tile.ts (hunk 16 above)
Comment (nikkimk) [reply elements/rh-tile/rh-tile.ts]: It's deliberately not a radio unless it is in a group that sets it as such because we don't want single radio tiles.

File: elements/rh-tile/rh-tile-group.css (hunk 12 above)
Comment (nikkimk) [reply elements/rh-tile/rh-tile-group.css:1]: @bennypowers There's an accessibility issue with using our own disabled as we want the tile to be readable by assistive tech just not actionable.

File: elements/rh-tile/rh-tile-group.ts [hunk 23]
```diff
@@ -138,7 +137,8 @@ export class RhTileGroup extends LitElement {
     this.#tiles = [...this.querySelectorAll('rh-tile')];
     this.#tiles.forEach(tile => {
       tile.checkable = true;
-      tile.radio = this.radio;
+      // @ts-expect-error: internal use of private prop. replace with context. see rh-tile.ts
+      tile.radioGroup = this.radio;
       tile.disabled !== this.disabled;
```
Comment (bennypowers) [inline elements/rh-tile/rh-tile-group.ts:142]: should be `tile.disabledGroup = this.disabled`?

**Review by zeroedin — APPROVED** [review]
Let's Get Tile Merged
